src/plcc/load_spec/validate_spec/validate_syntactic_spec/errors.py

```diff
@@ -2,7 +2,7 @@
 
 
 @dataclass
-class ValidationError:
+class ValidationError(Exception):
```

Comment:
Are ValidationErrors actually ever raised as an Exception through Python's exception handling mechanism? If not, I think we should not make them inherit from Exception.

Reply:
@StoneyJackson I made the ValidationError class inherit from Exception for testing purposes, as ```with raises``` expects a BaseException type, is it fine to leave it like this, or should I remove it? If I remove it, I am not sure how to test the errors, as there is no errorList for many of the sections, such as Grammar, which throws a few errors.

Reply:
We shouldn't extend Exception "for testing purposes". We should if we raise it as an exception.

When should we raise an exception? When the current algorithm cannot continue as normal because of the detected error.

Should a validator raise an exception? The purpose of a validator is to detect errors. If it is possible to detect multiple errors, it should do so. When it is done finding all the problems it can, it should return them as a list of errors. The only reason a validator would raise an exception is if there is something that would prevent it from doing its job.

Assuming a ValidationError represents an error detected by a validator (and not an error with the validator itself) it should not necessarily be an Exception.

One last thought. There is a chance later down the road that we actually do want to raise them as exceptions. But that's not clear at this point.

Reply:
@StoneyJackson ValidationErrors are extending Exceptions in different parts of the code as well, this may need to be added onto #38, and be dealt with in a bigger sense. But in the sense of this PR, there are errors that do not allow the LL1check to progress, would you like these extend Exception rather than ValidationError? Also, for the errors that don't prevent progress, would you like me to add an errorList that gets returned? I am thinking that the same errorList would go through each step and be extended, if needed, and that would be what is returned at the very end?

Reply:
I see... there should be a separation between a ValidationError and an exception that is raised when something breaks a validator.

I suggest that you rename all the exceptions with `Exception` at the end of their name, and have them extend `Exception` directly (and not `ValidationError`).

Then, for those errors that are not raised and represent errors detected by a validator, have them extend ValidatorError.

We'll worry about what the ValidatorError extends later.